src/app/views/canvas/mark_editor.tsx

```diff
@@ -221,6 +272,50 @@ export class SingleMarkView
     };
   }
 
+  public doZoomIn() {
+    const { scale, centerX, centerY } = this.state.zoom;
+    const fixPoint = Geometry.unapplyZoom(this.state.zoom, {
+      x: this.props.width / 2,
+      y: this.props.height / 2
+    });
+    let newScale = scale * 1.1;
+    newScale = Math.min(20, Math.max(0.05, newScale));
+    this.setState({
+      zoom: {
+        centerX: centerX + (scale - newScale) * fixPoint.x,
+        centerY: centerY + (scale - newScale) * fixPoint.y,
+        scale: newScale
+      }
+    });
+  }
+
+  public doZoomOut() {
+    const { scale, centerX, centerY } = this.state.zoom;
+    const fixPoint = Geometry.unapplyZoom(this.state.zoom, {
+      x: this.props.width / 2,
+      y: this.props.height / 2
+    });
+    let newScale = scale / 1.1;
+    newScale = Math.min(20, Math.max(0.05, newScale));
+    this.setState({
+      zoom: {
+        centerX: centerX + (scale - newScale) * fixPoint.x,
+        centerY: centerY + (scale - newScale) * fixPoint.y,
+        scale: newScale
+      }
+    });
+  }
```

Comment:
You can still have doZoomOut and in, but they are very similar, I would pull the logic out into a single function

```
public doZoom(factor: number) {
    ...
    let newScale = scale * factor;
    ...
}
```

then

```
public doZoomOut() {
    this.doZoom(.9);
}
public doZoomIn() {
   this.doZoom(1.1);
}
```